File: evals/refs/det.zig
const std = @import("std");
pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();
    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;
    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);
    var out_buf: [4096]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;
    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var c: usize = 0;
    while (c < t) : (c += 1) {
        const n = try std.fmt.parseInt(usize, it.next().?, 10);
        const m = try arena.alloc([]f64, n);
        for (m) |*row| row.* = try arena.alloc(f64, n);
        var i: usize = 0;
        while (i < n) : (i += 1) {
            var j: usize = 0;
            while (j < n) : (j += 1) m[i][j] = try std.fmt.parseFloat(f64, it.next().?);
        }
        var sign: f64 = 1.0;
        var zero = false;
        var col: usize = 0;
        while (col < n) : (col += 1) {
            var pivot = col;
            var r2 = col + 1;
            while (r2 < n) : (r2 += 1) {
                if (@abs(m[r2][col]) > @abs(m[pivot][col])) pivot = r2;
            }
            if (pivot != col) {
                const tmp = m[col]; m[col] = m[pivot]; m[pivot] = tmp;
                sign = -sign;
            }
            if (@abs(m[col][col]) < 1e-15) { zero = true; break; }
            r2 = col + 1;
            while (r2 < n) : (r2 += 1) {
                const factor = m[r2][col] / m[col][col];
                var c2 = col;
                while (c2 < n) : (c2 += 1) m[r2][c2] -= factor * m[col][c2];
            }
        }
        var det: f64 = if (zero) 0.0 else sign;
        if (!zero) {
            var k: usize = 0;
            while (k < n) : (k += 1) det *= m[k][k];
        }
        try w.print("{d:.6}\n", .{det});
    }
    try w.flush();
}

File: evals/refs/det.rs
use std::io::{self, Read, Write, BufWriter};

fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut it = s.split_ascii_whitespace();
    let t: usize = it.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let n: usize = it.next().unwrap().parse().unwrap();
        let mut m = vec![vec![0.0f64; n]; n];
        for i in 0..n {
            for j in 0..n {
                m[i][j] = it.next().unwrap().parse().unwrap();
            }
        }
        let mut sign = 1.0f64;
        let mut zero = false;
        for col in 0..n {
            let mut pivot = col;
            for r in (col+1)..n {
                if m[r][col].abs() > m[pivot][col].abs() { pivot = r; }
            }
            if pivot != col {
                m.swap(col, pivot);
                sign = -sign;
            }
            if m[col][col].abs() < 1e-15 { zero = true; break; }
            for r in (col+1)..n {
                let factor = m[r][col] / m[col][col];
                for c2 in col..n { m[r][c2] -= factor * m[col][c2]; }
            }
        }
        let det = if zero {
            0.0
        } else {
            let mut d = sign;
            for i in 0..n { d *= m[i][i]; }
            d
        };
        writeln!(out, "{:.6}", det).unwrap();
    }
}
